src/commonMain/kotlin/au/id/micolous/metrodroid/transit/serialonly/HoloTransitData.kt

```diff
@@ -0,0 +1,126 @@
+import au.id.micolous.metrodroid.multi.R
+import au.id.micolous.metrodroid.time.Epoch
+import au.id.micolous.metrodroid.time.MetroTimeZone
+import au.id.micolous.metrodroid.time.TimestampFormatter
+import au.id.micolous.metrodroid.transit.CardInfo
+import au.id.micolous.metrodroid.transit.TransitIdentity
+import au.id.micolous.metrodroid.transit.TransitRegion
+import au.id.micolous.metrodroid.ui.ListItem
+
+/**
+ * Transit data type for HOLO card.
+ *
+ *
+ * This is a very limited implementation of reading HOLO, because only
+ * little data is stored on the card
+ *
+ *
+ * Documentation of format: https://github.com/metrodroid/metrodroid/wiki/HOLO
+ */
+@Parcelize
+data class HoloTransitData(private val mSerial: Int?,
+                                private val mLastTransactionTimestamp: Int?, private val mManufacturingId: String?) : SerialOnlyTransitData() {
+
+    public override val extraInfo: List<ListItem>
+        get() = listOf(
+            ListItem(R.string.last_transaction, when (mLastTransactionTimestamp) {
+                0 -> "Never"
+                null -> "Never"
+                else -> TimestampFormatter.dateTimeFormat(parseTime(mLastTransactionTimestamp)).toString()
+            }),
+            ListItem(R.string.manufacture_id, mManufacturingId?.let { formatMfgId(mManufacturingId)})
+            )
+
+    override val reason: Reason
+        get() = Reason.NOT_STORED
+
+    override val cardName get() = NAME
+
+    override val serialNumber get() = formatSerial(mSerial)
+
+    companion object {
+        private const val NAME = "HOLO"
+        private const val APP_ID = 0x6013f2
+
+        private val TZ = MetroTimeZone.HONOLULU
+
+        private val CARD_INFO = CardInfo(
+                name = NAME,
+                cardType = CardType.MifareDesfire,
+                imageId = R.drawable.holo_card,
+                locationId = R.string.location_oahu,
+                region = TransitRegion.USA,
+                resourceExtraNote = R.string.card_note_card_number_only)
+
+        private fun parse(card: DesfireCard): HoloTransitData? {
+            val app = card.getApplication(APP_ID) ?: return null
+            val file1 = app.getFile(1)?.data
+            val serial = parseSerial(app)
+            val mfgId = (app.getFile(0)?.data?.convertBCDtoInteger(8, 3).toString() + app.getFile(0)?.data?.byteArrayToInt(0xb, 3).toString())
+            val lastTransactionTimestamp = file1?.byteArrayToInt(8, 4)
+            if (serial == null && lastTransactionTimestamp == null) return null
+
+            return HoloTransitData(
+                    mSerial = serial,
+                mLastTransactionTimestamp = lastTransactionTimestamp,
+                mManufacturingId = mfgId)
+        }
+
+        private fun parseSerial(app: DesfireApplication?) =
+                app?.getFile(0)?.data?.convertBCDtoInteger(0xe, 2)
+
+        val FACTORY: DesfireCardTransitFactory = object : DesfireCardTransitFactory {

```

Comment:
It's now better to use object syntax:
object HoloTransitFactory : DesfireCardTransitFactory {
...

Reply:
This style looks like it was copied from:

https://github.com/metrodroid/metrodroid/blob/288749e16511dfd7361ac35045ebaefea1c1193f/src/commonMain/kotlin/au/id/micolous/metrodroid/transit/serialonly/TrimetHopTransitData.kt#L94